From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <Blake.Galbreath@GMail.com>
Date: Sun, 23 Aug 2020 16:32:11 +0200
Subject: [PATCH] Add moon phase API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 8afcd707b03575f1040b8c9fab469a7958b9393d..2f1bbd89c02b2254e1437f1cc096c61b44c5a0f0 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -326,6 +326,11 @@ public class CraftWorld implements World {
     public int getPlayerCount() {
         return world.players.size();
     }
+
+    @Override
+    public io.papermc.paper.world.MoonPhase getMoonPhase() {
+        return io.papermc.paper.world.MoonPhase.getPhase(getFullTime() / 24000L);
+    }
     // Paper end
 
     private static final Random rand = new Random();
